axum-handle-error-extract/src/lib.rs

```diff
@@ -0,0 +1,389 @@
+//!             StatusCode::REQUEST_TIMEOUT,
+//!             format!("{} {} took too long", method, uri),
+//!         )
+//!     } else {
+//!         (
+//!             StatusCode::INTERNAL_SERVER_ERROR,
+//!             format!("{} {} failed: {}", method, uri, error),
+//!         )
+//!     }
+//! }
+//! # async {
+//! # axum::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+//! # };
+//! ```
+//!
+//! Not running any extractors is also supported:
+//!
+//! ```rust
+//! use axum::{
+//!     Router,
+//!     BoxError,
+//!     response::IntoResponse,
+//!     http::StatusCode,
+//!     routing::get,
+//! };
+//! use tower::{ServiceBuilder, timeout::error::Elapsed};
+//! use std::time::Duration;
+//! use axum_handle_error_extract::HandleErrorLayer;
+//!
+//! let app = Router::new()
+//!     .route("/", get(|| async {}))
+//!     .layer(
+//!         ServiceBuilder::new()
+//!             .layer(HandleErrorLayer::new(handle_error))
+//!             .timeout(Duration::from_secs(10))
+//!     );
+//!
+//! // this function just takes the error
+//! async fn handle_error(error: BoxError) -> impl IntoResponse {
+//!     if error.is::<Elapsed>() {
+//!         (
+//!             StatusCode::REQUEST_TIMEOUT,
+//!             "Request timeout".to_string(),
+//!         )
+//!     } else {
+//!         (
+//!             StatusCode::INTERNAL_SERVER_ERROR,
+//!             format!("Unhandled internal error: {}", error),
+//!         )
+//!     }
+//! }
+//! # async {
+//! # axum::Server::bind(&"".parse().unwrap()).serve(app.into_make_service()).await.unwrap();
+//! # };
+//! ```
+//!
+//! See [`axum::error_handling`] for more details on axum's error handling model and
+//! [`axum::extract`] for more details on extractors.
+//!
+//! # The future
+//!
+//! In axum 0.4 this will replace the current [`axum::error_handling::HandleErrorLayer`].
+
+#![warn(
+    clippy::all,
+    clippy::dbg_macro,
+    clippy::todo,
+    clippy::empty_enum,
+    clippy::enum_glob_use,
+    clippy::mem_forget,
+    clippy::unused_self,
+    clippy::filter_map_next,
+    clippy::needless_continue,
+    clippy::needless_borrow,
+    clippy::match_wildcard_for_single_variants,
+    clippy::if_let_mutex,
+    clippy::mismatched_target_os,
+    clippy::await_holding_lock,
+    clippy::match_on_vec_items,
+    clippy::imprecise_flops,
+    clippy::suboptimal_flops,
+    clippy::lossy_float_literal,
+    clippy::rest_pat_in_fully_bound_structs,
+    clippy::fn_params_excessive_bools,
+    clippy::exit,
+    clippy::inefficient_to_string,
+    clippy::linkedlist,
+    clippy::macro_use_imports,
+    clippy::option_option,
+    clippy::verbose_file_reads,
+    clippy::unnested_or_patterns,
+    rust_2018_idioms,
+    future_incompatible,
+    nonstandard_style,
+    missing_debug_implementations,
+    missing_docs
+)]
+#![deny(unreachable_pub, private_in_public)]
+#![allow(elided_lifetimes_in_paths, clippy::type_complexity)]
+#![forbid(unsafe_code)]
```

Comment:
FWIW: Assuming that these lint configurations are (mostly) the same as in axum, you can enable them for the entire workspace with [a small hack](Assuming that these lint configurations are (mostly) the same as in axum, you can enable them for the entire workspace with [a small hack](https://github.com/EmbarkStudios/rust-ecosystem/issues/22#issuecomment-947011395). The downside is that Rust <1.55 will choke on the necessary config, so an extra workaround (likely `rm -r .cargo` at the start of the MSRV CI job) would be required.

Reply:
That seems to also enable the lints for all the examples which is a little annoying 🤔

Reply:
Oh, right. Some might be useful for examples but most probably not.

I guees *if* you move all the library crates into a subdir at some point this will make more sense (since `.cargo` is directory structure based and doesn't have to be next to a `Cargo.toml`).